src/unit_cell/atom_type.cpp

```diff
@@ -737,6 +733,338 @@ Atom_type::read_input(std::string const& str__)
+        std::string bstr         = "PP_BETA." + std::to_string(i + 1);
+        pugi::xml_node beta_node = upf.child("PP_NONLOCAL").child(bstr.data());
+        auto beta_raw            = vec_from_str<double>(beta_node.child_value());
+
+        int nr = beta_node.attribute("cutoff_radius_index").as_int();
+        if (nr == 0) {
+            for (int j = beta_raw.size() - 1; j >= 0; j--) {
+                if (abs(beta_raw[j]) > 1.0e-80) {
+                    nr = j + 1;
+                    break;
+                }
+            }
+        }
+        if (nr == 0) {
+            nr = beta_raw.size();
+        }
+
+        std::vector<double> beta(nr);
+        std::copy(beta_raw.begin(), beta_raw.begin() + nr, beta.begin());
+
+        if (static_cast<int>(beta.size()) > num_mt_points()) {
+            std::stringstream s;
+            s << "wrong size of beta functions for atom type " << symbol_ << " (label: " << label_ << ")" << std::endl
+              << "size of beta radial functions in the file: " << beta.size() << std::endl
+              << "radial grid size: " << num_mt_points();
+            RTE_THROW(s);
+        }
+        int l = beta_node.attribute("angular_momentum").as_int();
+        if (spin_orbit_coupling_) {
+            // we encode the fact that the total angular momentum j = l
+            // -1/2 or l + 1/2 by changing the sign of l
+            std::string bstr       = "PP_RELBETA." + std::to_string(i + 1);
+            pugi::xml_node so_node = upf.child("PP_SPIN_ORB").child(bstr.data());
+
+            double j = so_node.attribute("jjj").as_double();
+            if (j < (double)l) {
+                l *= -1;
+            }
+        }
+        // add_beta_radial_function(l, beta);
+        if (spin_orbit_coupling_) {
+            if (l >= 0) {
+                add_beta_radial_function(angular_momentum(l, 1), beta);
+            } else {
+                add_beta_radial_function(angular_momentum(-l, -1), beta);
+            }
+        } else {
+            add_beta_radial_function(angular_momentum(l), beta);
+        }
+    }
+
+    mdarray<double, 2> d_mtrx({nbf, nbf});
+    d_mtrx.zero();
+    auto v = vec_from_str<double>(upf.child("PP_NONLOCAL").child("PP_DIJ").child_value(), 0.5);
+
+    for (int i = 0; i < nbf; i++) {
+        for (int j = 0; j < nbf; j++) {
+            d_mtrx(i, j) = v[j * nbf + i];
+        }
+    }
+    d_mtrx_ion(d_mtrx);
+
+    pugi::xml_node nl_node = upf.child("PP_NONLOCAL");
+    if (!nl_node.child("PP_AUGMENTATION").empty()) {
+        for (int i = 0; i < nbf; i++) {
+            std::string istr     = "PP_BETA." + std::to_string(i + 1);
+            pugi::xml_node inode = nl_node.child(istr.data());
+            int li               = inode.attribute("angular_momentum").as_int();
+
+            for (int j = i; j < nbf; j++) {
+                std::string jstr     = "PP_BETA." + std::to_string(j + 1);
+                pugi::xml_node jnode = nl_node.child(jstr.data());
```

Comment:
I think in order to compatible with GBRV pseudos, `nqf!=0` should also be supported here. And GBRV pseudos has q_ij rather than q_ijl, https://github.com/QEF/q-e/blob/57a97fe3ac56c6862ee083bf10d27505835c092a/upflib/read_upf_v1.f90#L65-L67

Reply:
As far as I can tell, GBRV pseudos are stored as UPF version 1 files. We are not aiming to write such a parser. If you are planing to convert to GBRV pseudos to v2, we could cover them as well. But then, I would assume they would follow the UPF v2 standard?

Reply:
> But then, I assume they would follow the UPF v2 standard?

Yes, but the UPF v2 standard is for the format, the non-zero `nqf` and `q_ij` are still the valid fields according to the UPF v2 schema and according to the notes provided in https://github.com/QEF/qeschemas/pull/11#issuecomment-1238139482

Reply:
As far as I remember V2 UPF pseudos introduced Q_ij^\ell(r) radial functions as a hard standard. Long term it is better to conver V1 GBRV to V2.

Reply:
https://gist.github.com/unkcpz/c3287268c84ba46b3366edb5537d65c7

Here is an example I convert the UPF.v1 GBRV to UPF.v2. In the QE, it will then be read by [read_upf_new.f90](https://github.com/QEF/q-e/blob/develop/upflib/read_upf_new.f90) instead of fallback read by [read_upf_v1.f90](https://github.com/QEF/q-e/blob/develop/upflib/read_upf_v1.f90). 
It is true from Giannozzi's note (https://github.com/QEF/qeschemas/pull/11#issuecomment-1238139482) that `qfcoef` and `rinner` are to be considered obsolescent. The plan for the SSSP v2 is to provide GBRV with converting it to UPFv2 format (let me know if this is useful, I need more test on such UPFv2 file I converted), I assume in SIRIUS you still can use it? But as @toxa81 pointed, maybe in the long term should push to have use uniformed expression.

Reply:
I personally think that having the whole SSSP library in UPF v2 format would be very useful, especially for people outside of QE. I also think that having a single uniform standard is important. For example, in this case, we would have to implement a special case for `qfcoef` and `rinner`, which in the end, is not very different from having 2 distinct parsers.

As far as I understand, it is possible to go from UPF v1 `PP_QIJ` to UPF v2 `PP_QIJL` seamlessly. For example, the `upf_to_json` tool of SIRIUS parses both [UPF v1](https://github.com/electronic-structure/SIRIUS/blob/bd9bb4581a9ee6939f89001b1336e29859ced84c/apps/upf/upf1_to_json.py#L247-L354) and [UPF v2](https://github.com/electronic-structure/SIRIUS/blob/bd9bb4581a9ee6939f89001b1336e29859ced84c/apps/upf/upf2_to_json.py#L119-L152) to the same JSON structure, which happens to be very close to that of UPF v2 (`PP_QIJL`).

Reply:
> For example, the upf_to_json tool of SIRIUS parses both [UPF v1](https://github.com/electronic-structure/SIRIUS/blob/bd9bb4581a9ee6939f89001b1336e29859ced84c/apps/upf/upf1_to_json.py#L247-L354) and [UPF v2](https://github.com/electronic-structure/SIRIUS/blob/bd9bb4581a9ee6939f89001b1336e29859ced84c/apps/upf/upf2_to_json.py#L119-L152) to the same JSON structure 

That is cool, I didn't know it actually convert not only the format but the expression as well. I'll talk to @simonpintarelli to see how it can be realize and I can probably add it to my v1 -> v2 converter.